src/crypto/tls/ech.go

```diff
@@ -0,0 +1,824 @@
+	if config.ECHEnabled {
+		hpkeSuite, err := echAssembleHpkeCipherSuite(dummyKemId, dummyKdfId, dummyAeadId)
+		if err != nil {
+			return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+		}
+
+		pk, err := hpkeSuite.KEM.Deserialize(echDummyPublicKey)
+		if err != nil {
+			return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+		}
+
+		var dummyEch echClient
+		dummyEch.handle.suite.kdfId = dummyKdfId
+		dummyEch.handle.suite.aeadId = dummyAeadId
+
+		dummyEch.handle.configId = make([]byte, dummyKdfOutputLen)
+		if _, err = io.ReadFull(config.rand(), dummyEch.handle.configId); err != nil {
+			return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+		}
+
+		dummyEch.handle.enc, _, err = hpke.SetupBaseS(hpkeSuite, config.rand(), pk, nil)
+		if err != nil {
+			return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+		}
+
+		dummyEch.payload = make([]byte, len(hello.marshal())+dummyAeadOverheadLen)
+		if _, err = io.ReadFull(config.rand(), dummyEch.payload); err != nil {
+			return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+		}
+
+		hello.encryptedClientHelloOffered = true
+		hello.encryptedClientHello = dummyEch.marshal()
+	}
+
+	// Bypass ECH.
+	return hello, nil, nil
+}
+
+func (c *Conn) echAcceptOrBypass(hello *clientHelloMsg) (*clientHelloMsg, error) {
+	config := c.config
+	echProvider := config.ServerECHProvider
+
+	// Determine if ECH can be accepted for this connection.
+	var echCanAccept bool
+	if echProvider != nil &&
+		config.ECHEnabled &&
+		config.maxSupportedVersion() >= VersionTLS13 &&
+		hello.encryptedClientHelloOffered {
+		publicNames, err := echProvider.GetPublicNames()
+		if err != nil {
+			c.sendAlert(alertInternalError)
+			return nil, fmt.Errorf("ech: could not resolve public names: %s", err)
+		}
+
+		for _, name := range publicNames {
+			if hello.serverName == name {
+				echCanAccept = true
+			}
+		}
+	}
+
+	// Decide whether to bypass ECH.
+	if !echCanAccept {
+		if c.hrrTriggered && c.ech.accepted {
+			c.sendAlert(alertIllegalParameter)
+			return nil, fmt.Errorf("ech: hrr: bypass after accept")
+		}
+
+		// Bypass ECH.
+		return hello, nil
+	}
+
+	// Attempt to accept ECH.
+	//
+	// Parse the payload of the ECH extension.
+	rawClientEch := hello.encryptedClientHello
+	clientEch, err := echUnmarshalClient(rawClientEch, extensionECH)
+	if err != nil {
+		c.sendAlert(alertIllegalParameter)
+		return nil, fmt.Errorf("ech: %s", err)
+	}
+
+	// Ask the ECH provider for the decryption context. Note that c.ech.hrrPsk
+	// is initially nil, meaning no PSK is used for encrypting the first
+	// ClientHelloInner.
+	res, err := echProvider.GetServerContext(clientEch.handle.marshal(), c.ech.hrrPsk, extensionECH)
+	if err != nil {
+		// This condition indicates the connection must be aborted.
+		c.sendAlert(alert(res.Alert))
+		return nil, fmt.Errorf("ech: %s", err)
+	}
+
+	if res.Rejected {
+		if c.hrrTriggered && c.ech.accepted {
```

Comment:
Good catch! This made me realize that the server's behavior here is incorrect. The client-side errors that need to be enforced are bypass-after-offer and offer-after-bypass. This PR fixes this in the spec: https://github.com/tlswg/draft-ietf-tls-esni/pull/311. I also fixed it here.

Reply:
I was not concerned about the client behavior here, I assumed that the client uses the same ECHConfig for the both the first and post-HRR CH.

Code-wide, this currently happens on the server side:
1. `echProvider.GetServerContext` is called and succeeds, resulting in ECH acceptance. `c.ech.st.Offered = true` and `c.ech.st.Accepted = true` are set.
2. `doHelloRetryRequest` is invoked, so HRR is triggered. `echAcceptOrBypass` is called once again.
3. Consider that this time `echProvider.GetServerContext` **fails** because the `config_id` is no longer known. This could happen after key rotation just right after step 1.

After (3), `c.ech.retryConfigs` will also be set while `c.ech.st.Accepted` was already set in (1), This is also an internal consistency.

Reply:
The PR I listed also clarifies the server behavior: see https://github.com/tlswg/draft-ietf-tls-esni/pull/311/files#diff-4d2dc9df336bea8e17f5eb4ed7cb1107R765.

You're correct about the program flow, and about the inconsistency after rejection. I modified the server's behavior to match https://github.com/tlswg/draft-ietf-tls-esni/pull/311, which fixes the inconsistency.

One minor nit:

>     1. `echProvider.GetServerContext` is called and succeeds, resulting in ECH acceptance. `c.ech.st.Offered = true` and `c.ech.st.Accepted = true` are set.

ECH accept doesn't happen until after decryption, which happens after this call. It's true that `c.ech.st.Offered` is set at this point, but `c.ech.st.Accepted` is not set until all ECH processing is done.

Reply:
The linked tlswg PR is unrelated to the inconsistency concern described here. That PR tries to address a case where the client sends conflicting signals. Here I try to describe the case where the server loses the decryption key before processing the second (post-HRR) CH.

And sure, with (1) I meant that the server performs all steps leading up to ECH acceptance which implies successful decryption and availability of keys from `echProvider.GetServerContext`.

If you want I can try to craft an ECHProvider implementation that demonstrates this case (success on the first call, failure on the second).

Reply:
> The linked tlswg PR is unrelated to the inconsistency concern described here. That PR tries to address a case where the client sends conflicting signals. Here I try to describe the case where the server loses the decryption key before processing the second (post-HRR) CH.

I'm not sure we're understanding each other. Let me try to lay out the bug and how it was fixed. Previously, the server checked for HRR consistency (i.e., reject-after-accept) just before rejecting:
```
	if res.Rejected {
		if c.hrrTriggered && c.ech.accepted {
                     // Abort with "illegal_parameter" alert
                }
                // Reject ECH
         }
```
As you correctly pointed out, if key rotation happens after producing HRR but prior to consuming the second CH, then the server will abort when in fact it should have rejected. This observation made me realize that the HRR consistency check isn't correct. In the spec, I revised the client-facing server behavior to check for bypass-after-offer and offer-after-bypass only. For convenience, here's the key text from the open PR:

> It is an error for the client to offer ECH before the HelloRetryRequest but not
after. Likewise, it is an error for the client to offer ECH after the
HelloRetryRequest but not before. If either of these conditions occurs, then the
client-facing server MUST abort the handshake with an "illegal_parameter" alert.

This effectively eliminates the need for to check for reject-after-accept prior to rejection. Now consider what happens if key rotation happens after producing HRR but before consuming the second CH: the consistency check passes (no offer-after-bypass detected), so the server will reject. This is the correct behavior.

> If you want I can try to craft an ECHProvider implementation that demonstrates this case (success on the first call, failure on the second).

By all means! The more test cases the better :)

Reply:
See https://github.com/cloudflare/go/pull/30/commits/811744084cf4e0d9223880060a04a43b8f662f5d for the new test that tries to demo the issue.

Reply:
The server-side error (`"detected illegal second ClientHello"`) points to another issue. Because we're rejecting ECH this time, the CH consumed by the server has changed from the CHInner to the CHOuter. This is illegal because the CHInner parameters are different than the CHOuter parameters --- namely, the supported versions have changed --- which means HRR-sensitive parameters changed from CH1 to CH2. Hence, the server aborts with "illegal_parameter". But this abort isn't informative for the client, since it hasn't done anything wrong.

We could try to recover and reject properly. However, this could trigger *yet another HRR*, since the client-facing server is now terminating the connection instead of the backend server. Instead, I think the behavior should be that, if the server needs to reject after it accepted, then it aborts with "handshake_failure". The server also needs to ensure that the "config_id" and "suites" haven't changed across the HRR, which would trigger the same condition but would be the client's fault.

Damn, Peter --- really great work. This was subtle. I'm merging your test and pushing a fix.